### PR TITLE
Mob Detection for Chaos Sanctuary Hammerdin

### PR DESCRIPTION
**Pull request by jobithu**
- added cv2 functionality for basic mob detection 
- checks for green or blue colored-mobs in sufficient size before killing trash
- requires you to have a merc with either holyfreeze (blue), conviction (green) or andariel's (green)
- decreases run duration of CS by up to 30%
- sometimes gets distracted by a poisoned merc.

### CODE REVIEW

**Review by VladimirMakaev — COMMENTED** [review]
@jobithu Can we make it as a standalone package instead of putting everything in template finder? Also the PR title mentions Hammerdin, but it should be general feature, isn't it? Can we land a generic monster detection package first and then a follow up PR would be integration on hammerdin for CS, Trav, etc?

**Comment by jobithu** [comment]
> @jobithu Can we make it as a standalone package instead of putting everything in template finder? Also the PR title mentions Hammerdin, but it should be general feature, isn't it? Can we land a generic monster detection package first and then a follow up PR would be integration on hammerdin for CS, Trav, etc?

Refactored to separate file mob_detection.py. You can now call it for each char & each run.
Two downsides: redemption is not cast. and the mob detection parameters are hardcoded - on the other hand apart from blue/white Filtering, I do not know what makes sense.

I kept Hammerdin in, to give a proof of concept.

**Review by mgleed — COMMENTED** [review]
This looks like a good start. Give me some time to finish town overhaul and I'll join in on this.

File: src/mob_detection.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+# adapted from BEN Open CV Bot Tutorial #9
+def get_targets_ordered_by_distance(targets, ignore_radius:int=0):
+    my_pos = (1280/2, 720/2) #center of the screen
+    def pythagorean_distance(pos):
+        return sqrt((pos[0] - my_pos[0])**2 + (pos[1] - my_pos[1])**2)
+    targets.sort(key=pythagorean_distance)
+    targets = [t for t in targets if pythagorean_distance(t) > ignore_radius] #ignore targets that are too close
+    return targets #a sorted arry of all targets, nearest to farest away
+
+def mobcheck(self, info_ss:bool=False) -> bool:
```
[the remark this answers is not in the document]
Comment (VladimirMakaev) [reply src/mob_detection.py:149]: So can we change the public api to this package to something along the lines I posted above? From caller perspective we want to be super clear what's going on and to know as little as possible about the implementation details. I understand that in CV there could be a false positive and probably a caller needs to deal with it. If we can provide some "score" to them good if not then not.

File: src/mob_detection.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    #wait(1) # let the merc paint some mobs
+    img = grab()
```
[the remark this answers is not in the document]
Comment (jobithu) [reply src/mob_detection.py:151]: Sure, would then be grabbed in the line before the function is called.

**Comment by mgleed** [comment]
Looks like there are some errors that are causing the github unittests to fail.

i see one, I'm guessing this was a mistake when merging master?

image

**Comment by jobithu** [comment]
works well, but skips mobs occasionally

**Comment by mgleed** [comment]
> works well, but skips mobs occasionally

didn't you want it as a param before merging?

**Review by mgleed — APPROVED** [review]
I approve

**Pull request merged**